utils_nlp/models/xlnet/sequence_classification.py

```diff
@@ -0,0 +1,363 @@
+        labels,
+        val_token_ids,
+        val_input_mask,
+        val_labels,
+        token_type_ids=None,
+        val_token_type_ids=None,
+        verbose=True,
+        logging_steps=0,
+        save_steps=0,
+        val_steps=0,
+    ):
+        """Fine-tunes the XLNet classifier using the given training data.
+
+        Args:
+            token_ids (list): List of training token id lists.
+            input_mask (list): List of input mask lists.
+            labels (list): List of training labels.
+            token_type_ids (list, optional): List of lists. Each sublist
+                contains segment ids indicating if the token belongs to
+                the first sentence(0) or second sentence(1). Only needed
+                for two-sentence tasks.
+            verbose (bool, optional): If True, shows the training progress and
+                loss values. Defaults to True.
+        """
+
+        device = get_device("cpu" if self.num_gpus == 0 or not torch.cuda.is_available() else "gpu")
+        self.model = move_to_device(self.model, device, self.num_gpus)
+
+        token_ids_tensor = torch.tensor(token_ids, dtype=torch.long)
+        input_mask_tensor = torch.tensor(input_mask, dtype=torch.long)
+        labels_tensor = torch.tensor(labels, dtype=torch.long)
+
+        val_token_ids_tensor = torch.tensor(val_token_ids, dtype=torch.long)
+        val_input_mask_tensor = torch.tensor(val_input_mask, dtype=torch.long)
+        val_labels_tensor = torch.tensor(val_labels, dtype=torch.long)
+
+        if token_type_ids:
+            token_type_ids_tensor = torch.tensor(token_type_ids, dtype=torch.long)
+            val_token_type_ids_tensor = torch.tensor(val_token_type_ids, dtype=torch.long)
+
+            train_dataset = TensorDataset(
+                token_ids_tensor, input_mask_tensor, token_type_ids_tensor, labels_tensor
+            )
+
+            val_dataset = TensorDataset(
+                val_token_ids_tensor,
+                val_input_mask_tensor,
+                val_token_type_ids_tensor,
+                val_labels_tensor,
+            )
+
+        else:
+
+            train_dataset = TensorDataset(token_ids_tensor, input_mask_tensor, labels_tensor)
+
+            val_dataset = TensorDataset(
+                val_token_ids_tensor, val_input_mask_tensor, val_labels_tensor
+            )
+
+        # define optimizer and model parameters
+        param_optimizer = list(self.model.named_parameters())
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)],
+                "weight_decay": self.weight_decay,
+            },
+            {
+                "params": [p for n, p in param_optimizer if any(nd in n for nd in no_decay)],
+                "weight_decay": 0.0,
+            },
+        ]
+
+        train_sampler = RandomSampler(train_dataset)
```

Comment:
these two lines are repeated below, are you sure you need them?

Reply:
that can be cleaned up - will do it. thanks!